spark/src/main/java/org/apache/iceberg/spark/source/StructInternalRow.java

```diff
@@ -146,8 +146,13 @@ public UTF8String getUTF8String(int ordinal) {
 
```

Comment:
The change to this class was mostly trying to use it in `TestSparkParquetMergingMetrics`. Currently this class is only being used for reading rows for metadata tables (in [RowDataReader](https://github.com/apache/iceberg/blob/master/spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java#L189), and I think only the metadata table will produce `DataTask`). 

In the tests I wanted to convert `Record` to `InternalRow` for testing Spark appender. I was debating if I should expand this class beyond its current usage, or to write a new converter. Here are the two things I need to change to (partially) implement the former: 
1) null handling which results to the change in `get()`: `RowDataReader` doesn't call `get()` directly (it uses Dyn reflections for reading individual attributes and skip nulls) for converting into other Spark internal row representation (`UnsafeRow` in this case), thus we didn't see issue. However, when spark parquet writer uses `get()`, without this change NPE will occur. Note that even after this change, other use cases of this class (e.g. `getUTF8String()` are still not null safe, and I wonder if people have opinion on if we want a full null-safe support update to all methods in this class. 
2) for `getBinary()` change, currently we convert `Fixed` type to binary for Spark ([link](https://github.com/apache/iceberg/blob/master/spark/src/main/java/org/apache/iceberg/spark/TypeToSparkType.java#L118-L119)), and the method I used for creating random records generate fixed type with `byte[]` ([link](https://github.com/apache/iceberg/blob/master/api/src/test/java/org/apache/iceberg/util/RandomUtil.java#L130-L133)), and before this change `getBinary()` doesn't work for byte[] implementation of fixed type. Alternatively we could wrap fixed type in random record generator the [same way we do for binary type](https://github.com/apache/iceberg/blob/master/data/src/test/java/org/apache/iceberg/data/RandomGenericData.java#L229-L230). I decide to do the former to allow binary related types to be more flexible when wrapping them in this class, but I guess this comes back to the question to if we want to evolve this class or to create a separate converter.

Reply:
For #1, I'm fine adding the null check since this isn't used in a high performance code path, but it would be better to have the code that uses this call `isNullAt` directly because that is the contract for Spark's `InternalRow`.

For #2, since `getBinary` is called for both fixed and binary types, I think we do need to check the type of the object. I'd much rather do that using `instanceof` rather than catching an exception. Can you update it to use `struct.get(ordinal, Object.class)` and then check the type of the object returned?

Reply:
Thanks for the information!

For 1, thanks for the info! I wasn't aware of the contract. 

I wasn't sure if we want to add `isNullAt` for this specific case though, as I guess the problem comes from a difference in behavior between `InternalRow` and `StructInternalRow`, and adding `isNullAt` might have a larger performance penalty in production.  

The behavior difference comes from the ability of calling `internalRow.get()` that could return null. The NPE eventually comes from [`struct.get(index, types[index])`](https://github.com/apache/iceberg/blob/master/spark/src/main/java/org/apache/iceberg/spark/data/SparkParquetWriters.java#L420) in `SparkParquetWriters`. While the actual `InternalRow` could return null for null column, for `StructInternalRow` it assumes the underlying `get()` returns non-null, and sometimes performs some actions to them that could lead to NPE. e.g. [call `toString` for converting](https://github.com/apache/iceberg/blob/master/spark/src/main/java/org/apache/iceberg/spark/source/StructInternalRow.java#L144), or [return as primitive type directly](https://github.com/apache/iceberg/blob/master/spark/src/main/java/org/apache/iceberg/spark/source/StructInternalRow.java#L132). Thus `SparkParquetWriters` was able to call `get` fine under normal circumstances, but it couldn't for the specific usage of `StructInternalRow` in this test. 

For 2, that's a better idea, will do this instead.

Reply:
For 1, let's just add the check. This isn't used in a performance-critical path.

Reply:
This looks good to me.